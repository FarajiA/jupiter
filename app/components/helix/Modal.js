import PropTypes from 'prop-types';
import React, { Component } from 'react';

import Header from './modal/Header';
import Body from './modal/Body';
import Footer from './modal/Footer';

/**
 * Helix Modal
 * @example https://rackerlabs.github.io/helix-ui/components/modals/
 */
class Modal extends Component {
  componentDidMount() {
    if (this.hxModal) {
      this.hxModal.addEventListener('close', this.props.onClose);
      this.hxModal.addEventListener('open', this.props.onOpen);
    }
  }

  componentWillUnmount() {
    if (this.hxModal) {
      this.hxModal.removeEventListener('close', this.props.onClose);
      this.hxModal.removeEventListener('open', this.props.onOpen);
    }
  }

  setRef = (element) => { this.hxModal = element; };

  render() {
    return (
      <hx-modal
        id={this.props.id}
        class={`Modal ${this.props.className} ${this.props.size}`}
        ref={this.setRef}
        open={this.props.isOpen ? true : null}
      >
        {this.props.children}
      </hx-modal>
    );
  }
}

Modal.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  isOpen: PropTypes.bool,
  onClose: PropTypes.func,
  size: PropTypes.string,
  onOpen: PropTypes.func
};

Modal.defaultProps = {
  size: 'medium',
  isOpen: false,
  className: ''
};

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

export default Modal;
