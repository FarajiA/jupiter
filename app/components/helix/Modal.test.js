import React from 'react';
import * as enzyme from 'enzyme';
import Modal from './Modal';
import renderer from 'react-test-renderer';

describe('Modal', () => {
  const defaultProps = { id: 'fooId' };
  const onCloseMock = jest.fn();
  const onOpenMock = jest.fn();
  const defaultMountProps = { onClose: onCloseMock, onOpen: onOpenMock };
  const defaultContent = <p>Modal content</p>;
  const shallow = (props = defaultProps, content = defaultContent) => {
    return enzyme.shallow(<Modal {...props}>{content}</Modal>).find('hx-modal');
  };
  const mount = (props = defaultProps, content = defaultContent) => {
    return enzyme.mount(<Modal {...props} {...defaultMountProps}>{content}</Modal>);
  };

  it('renders with props and default props', () => {
    const rendered = renderer.create(<Modal {...defaultProps} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders children', () => {
    expect(shallow().prop('children')).toEqual(<p>Modal content</p>);
  });

  it('renders with className prop', () => {
    const modal = shallow({ className: 'modalClass', ...defaultProps });
    expect(modal.prop('class')).toEqual('Modal modalClass medium');
  });

  it('Sets modal open attribute to true', () => {
    const modal = shallow({ isOpen: true, ...defaultProps });
    expect(modal.prop('open')).toBeTruthy();
  });

  it('does not set modal open attribute to false', () => {
    const modal = shallow({ isOpen: false, ...defaultProps });
    expect(modal.prop('open')).toEqual(null);
  });

  it('calls onClose when closed', () => {
    const modal = mount();
    modal.prop('onClose')();
    expect(onCloseMock).toBeCalledTimes(1);
  });

  it('componentDidMount adds event listeners', () => {
    const instance = mount().instance();
    instance.hxModal.addEventListener = jest.fn();
    instance.componentDidMount();
    expect(instance.hxModal.addEventListener).toHaveBeenCalledTimes(2);
    expect(instance.hxModal.addEventListener).toHaveBeenCalledWith('open', onOpenMock);
    expect(instance.hxModal.addEventListener).toHaveBeenCalledWith('close', onCloseMock);
  });

  it('componentWillUnmount removes all event listeners', () => {
    const instance = mount().instance();
    instance.hxModal.removeEventListener = jest.fn();
    instance.componentWillUnmount();
    expect(instance.hxModal.removeEventListener).toHaveBeenCalledTimes(2);
    expect(instance.hxModal.removeEventListener).toHaveBeenCalledWith('open', onOpenMock);
    expect(instance.hxModal.removeEventListener).toHaveBeenCalledWith('close', onCloseMock);
  });

  describe('size', () => {
    it('renders large', () => {
      const modal = shallow({ size: 'large' });
      expect(modal.prop('class')).toEqual('Modal  large');
    });

    it('renders medium', () => {
      const modal = shallow({ size: 'medium' });
      expect(modal.prop('class')).toEqual('Modal  medium');
    });

    it('renders small', () => {
      const modal = shallow({ size: 'small' });
      expect(modal.prop('class')).toEqual('Modal  small');
    });
  });
});
