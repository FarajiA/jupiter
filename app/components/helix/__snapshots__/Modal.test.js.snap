// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders with props and default props 1`] = `
<hx-modal
  class="Modal  medium"
  id="fooId"
  open={null}
/>
`;
